ir/memory.cpp

```diff
@@ -815,11 +811,29 @@ expr Pointer::blockValRefined(const Pointer &other) const {
                 val.ptrByteoffset() == val2.ptrByteoffset() &&
                 val.ptr().refined(val2.ptr());
   }
+  // allow null ptr <-> zero
+  // If ptr points to a null block with ofs, its integer address is compared.
+  // Relevant transformations:
+  //   store ty* null  <-> memset i8 0
+  //   store ty* undef  -> memset i8 undef
+  expr intval;
+  if (val.isValid()) {
+    expr ofs = val.ptr().getOffset();
+    expr shift_amount = val.ptrByteoffset() *
+                        expr::mkUInt(bits_byte, val.ptrByteoffset().bits());
+    intval = ofs.lshr(shift_amount.zextOrTrunc(ofs.bits()))
```

Comment:
I don't really understand this.
You have `intval = offset >> byteOffset * bits_byte`

This formula has no meaning. It's mixing ptr offset & the byte offset that can't really be mixed. What's this trying to achieve?

Reply:
The goal here is to interpret a pointer byte (0, ofs) as a 64-bit integer of ofs.
When loading a byte, the corresponding 8 bits should be extracted, so it is shifted by `byteOffset * bits_byte` and extracted.

Reply:
This doesn't seem correct. See this example:
```
p = malloc
p' = gep p, 0x700000
store i8* p, i8** q
=>
store i8* p, i8** q
store 0, (bitcast p to i8)
```

Your code will shift the offset by at least one. So the shifted offset is 0, and thus refinement will hold for the above program. Which it shouldn't.

Reply:
I think in that case p isn't null block, so refinement shouldn't hold.
Here is the program that I tested with - [zero-null2.zip](https://github.com/AliveToolkit/alive2/files/4640248/zero-null2.zip)

Reply:
ok, then this:
```
p' = gep null, 0x700...000
store i8* p, i8** q
=>
store i8* null, i8** q
store 0, (bitcast p to i8)
```

I really don't understand the direction you are taking with this patch. It compares 2 completely different types: internal Alive2 pointer representation with an `i8` LLVM value. Doesn't type check. Can't see why it would ever be correct.

Reply:
In that case refinement also doesn't hold, if I understand the example correctly: https://gist.github.com/aqjune/d3d76604f28a71f8e4af565856b83eac

There are two optimizations that fails if we allow refinement on bytes of a same type only:
`store * null, %ptr   -> memset i8 0` (#236)
`store * undef, %ptr -> memset i8 undef` (#345)
I was trying to give a relation between a pointer byte and i8 to explain these. The first one was addressed by #237, the second one is on the fly.
An approach that I suggest is that we can treat a pointer on a null block as something special so it is compatible with an integer. I think this is a natural expansion of the previous patch (#237) as well as supports these two optimizations.